File: native-spring-api/src/main/java/com/xavierbouclet/nativeapi/MovieService.java
package com.xavierbouclet.nativeapi;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.UUID;

import static java.util.Set.copyOf;

record MovieService(
        MovieRepository repository,
        ActorRepository actorRepository,
        ActorMovieRepository actorMovieRepository
) {

    public Flux<Movie> list() {
        return repository.findAll().flatMap(movie -> Mono.just(movie).zipWith(actorRepository.findActorsByMovieId(movie.getId()).collectList())
                .map(result -> {
                    result.getT1().setActors(copyOf(result.getT2()));
                    return result.getT1();
                }));
    }

    public Mono<Movie> add(Movie movie) {
        return repository.save(movie);
    }

    public Mono<Movie> modify(Movie movie) {
        return repository.save(movie);
    }

    public Mono<UUID> delete(UUID id) {
        return repository.deleteById(id).map(v -> id);
    }

    public Mono<Movie> get(UUID id) {
        return repository.findById(id).zipWith(actorRepository.findActorsByMovieId(id).collectList())
                .map(result -> {
                    result.getT1().setActors(copyOf(result.getT2()));
                    return result.getT1();
                });
    }

    public Mono<ActorMovie> addActorToMovie(UUID id, UUID actorId) {
        return actorMovieRepository.save(new ActorMovie(actorId,id));
    }

    public Mono<Void> removeActorFromMovie(UUID id, UUID actorId) {
        return actorMovieRepository.deleteByMovieIdAndActorId(id, actorId);
    }

}

File: native-spring-api/src/main/kotlin/com/xavierbouclet/nativeapi/MovieService.kt
package com.xavierbouclet.nativeapi

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import java.util.*

@Service
class MovieService(
    @Autowired val repository: MovieRepository,
    @Autowired val actorRepository: ActorRepository
) {

    fun list(): MutableIterable<Movie> {
        return repository.findAll()
    }

//    fun list(title: String): List<Movie> {
//        return repository.findByTitle(title)
//    }

    fun add(movie: Movie): Movie {
        return repository.save(movie)
    }

    fun modify(movie: Movie): Movie {
        return repository.save(movie)
    }

    fun delete(id: UUID): UUID {
        repository.deleteById(id)
        return id
    }

    fun addActorToMovie(id: UUID, idActor: UUID): Movie {
        val movie = repository.findById(id).get()
        val actor = actorRepository.findById(idActor).get()
        movie.actors?.add(actor)
        actor.movies?.add(movie)
        return repository.save(movie)
    }

    fun get(id: UUID): Movie? {
        return repository.findByIdOrNull(id)
    }

    fun removeActorFromMovie(id: UUID, idActor: UUID): Movie {
        val movie = repository.findById(id).get()
        val actor = actorRepository.findById(idActor).get()
        movie.actors?.remove(actor)
        actor.movies?.remove(movie)
        return repository.save(movie)
    }

}
